src/sync/sync_permission.cpp

```diff
@@ -157,7 +157,91 @@ void Permissions::set_permission(std::shared_ptr<SyncUser> user,
+        {"mayManage", permission.access == AccessLevel::Admin},
+    };
+    switch (permission.condition.type) {
+        case Permission::Condition::Type::KeyValue:
+            props.insert({"metadataKey", permission.condition.key_value.first});
+            props.insert({"metadataValue", permission.condition.key_value.second});
+            break;
+        default:
+            break;
+    }
+    auto cb = [callback=std::move(callback)](Object*, std::exception_ptr exception) {
+        if (exception) {
+            callback(exception);
+        } else {
+            callback(nullptr);
+        }
+    };
+    perform_async_operation("PermissionChange", std::move(user), std::move(cb), std::move(props), make_config);
+}
+
+void Permissions::delete_permission(std::shared_ptr<SyncUser> user,
+                                    Permission permission,
+                                    PermissionChangeCallback callback,
+                                    const ConfigMaker& make_config)
+{
+    permission.access = AccessLevel::None;
+    set_permission(std::move(user), std::move(permission), std::move(callback), make_config);
+}
+
+void Permissions::make_offer(std::shared_ptr<SyncUser> user,
+                             PermissionOffer offer,
+                             PermissionOfferCallback callback,
+                             const ConfigMaker& make_config)
+{
+    auto props = AnyDict{
+        {"expiresAt", std::move(offer.expiration)},
+        {"userId", user->identity()},
+        {"realmUrl", user->server_url() + offer.path},
+        {"mayRead", offer.access != AccessLevel::None},
+        {"mayWrite", offer.access == AccessLevel::Write || offer.access == AccessLevel::Admin},
+        {"mayManage", offer.access == AccessLevel::Admin},
+    };
+    auto cb = [callback=std::move(callback)](Object* object, std::exception_ptr exception) {
+        if (exception) {
+            callback(none, exception);
+        } else {
+            CppContext context;
+            auto token = any_cast<std::string>(object->get_property_value<util::Any>(context, "token"));
+            callback(util::make_optional<std::string>(std::move(token)), nullptr);
+        }
+    };
+    perform_async_operation("PermissionOffer", std::move(user), std::move(cb), std::move(props), make_config);
+}
+
+void Permissions::accept_offer(std::shared_ptr<SyncUser> user,
+                               const std::string& token,
+                               PermissionOfferCallback callback,
+                               const ConfigMaker& make_config)
+{
+    auto props = AnyDict{
+        {"token", token},
+    };
+    auto cb = [callback=std::move(callback)](Object* object, std::exception_ptr exception) {
+        if (exception) {
+            callback(none, exception);
+        } else {
+            CppContext context;
+            auto token = any_cast<std::string>(object->get_property_value<util::Any>(context, "realmUrl"));
```

Comment:
The string here is the only thing that differs from the above, so it's probably worth extracting this all to a function.